src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/BootstrapMessage.java

```diff
@@ -0,0 +1,94 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.cloud.spanner.pgadapter.wireprotocol;
+
+import com.google.cloud.spanner.pgadapter.ConnectionHandler;
+import com.google.cloud.spanner.pgadapter.wireoutput.AuthenticationOkResponse;
+import com.google.cloud.spanner.pgadapter.wireoutput.KeyDataResponse;
+import com.google.cloud.spanner.pgadapter.wireoutput.ReadyResponse;
+import com.google.cloud.spanner.pgadapter.wireoutput.ReadyResponse.Status;
+import com.google.cloud.spanner.pgadapter.wireoutput.StartUpMessageResponse;
+import java.io.DataOutputStream;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * This represents all messages which occur before {@link ControlMessage} type messages. Those
+ * include encryption, admin (e.g.: cancellation) and start-up messages.
+ */
+public abstract class BootstrapMessage extends WireMessage {
+
+  public BootstrapMessage(ConnectionHandler connection, int length) {
+    super(connection, length);
+  }
+
+  /**
+   * Factory method to create the bootstrap message from their designated format. Note this
+   * is backwards from control messages where identifier is first and length second.
+   *
+   * @param connection The connection handler object setup with the ability to send/receive.
+   * @return The constructed wire message given the input message.
+   * @throws Exception If construction or reading fails.
+   */
+  public static BootstrapMessage create(ConnectionHandler connection)
+      throws Exception {
+    int length = connection.getConnectionMetadata().getInputStream().readInt();
+    int protocol = connection.getConnectionMetadata().getInputStream().readInt();
+    switch (protocol) {
+      case SSLMessage.IDENTIFIER:
+        return new SSLMessage(connection);
+      case StartupMessage.IDENTIFIER:
+        return new StartupMessage(connection, length);
+      case CancelMessage.IDENTIFIER:
+        return new CancelMessage(connection);
+      default:
+        throw new IllegalStateException("Unknown message");
+    }
+  }
+
+  /**
+   * Parses parameters specific to bootstrap messages. Those generally (unlike control parameters)
+   * exclude metadata including length and are simple null (0) delimited.
+   *
+   * @param rawParameters The input string containing parameters (null delimited)
+   *
+   * @return A KV map of the parameters
+   */
+  protected Map<String, String> parseParameters(String rawParameters) {
+    Map<String, String> parameters = new HashMap<>();
+    String[] paramArray = rawParameters.split(new String(new byte[]{(byte) 0}));
+    for (int i = 0; i < paramArray.length; i = i + 2) {
```

Comment:
We expect them to be properly formatted; unfortunately the protocol itself makes this a very light start-up, which means there is little we can do to protect against that (we don't even get string length or number of params, like regular, non-startup message), so it's hard to protect against that. That being said, we do basically ignore these parameters currently, so even if improperly formatted, there is zero risk.

Reply:
my point was that 1 extra entry in this list would be silently dropped and nobody could notice.  Perhaps a log message if there's a mismatch?

Reply:
As it stands, if the number isn't even, an ArrayIndexOutOfBoundsException will be thrown, which will be logged and the surrounding logs will say what the message was since we log the input payloads, so we should be good on that front :)